contributor-docs/adrs/adr-016-css.md

```diff
@@ -0,0 +1,103 @@
+# ADR 16: Styling with CSS
+
+## Status
+
+| Stage    | Status                        |
+| -------- | ----------------------------- |
+| Approved | ✅                            |
+| Adopted  | 🚧 _TODO: add tracking issue_ |
+
+## Context
+
+Reasons for change: Performance, utilise css variables
+
+The main reason for changing our styling architecture is to remove performance issues for applications:
+
+Relevant parts from [primer/discussions#1928](https://github.com/github/primer/discussions/1928#reasons)(Hubbers only link):
+
+1. Initial page loads take longer
+
+   Because styles are injected during runtime, we can not parallelize this work. The page is un-styled till the javascript bundle is
+   downloaded and processed. The processing time is longer on slower devices.
+
+   Benchmarks:
+
+   i. Lab benchmarks: rendering a page with 1000 components
+   runtime injection of css: 242ms
+   static .css files: 96ms (60% faster)
+
+   ii. Application benchmarks: render times for repo directory (>1000 files, [gh/gh#241090](https://github.com/github/github/pull/241090))
+   with primer/react Box: 3800ms
+   with custom css: 3076ms (20% faster)
+
+2. Slow server side rendering
+
+   Some applications are able to make up for slower initial render times by collecting styles on the server first. In the correct stack, collecting styles requires rendering the app twice before we can ship any html to the user.
+
+   benchmarks: in the issues-app experiment, collecting styles on the server contributed to + 20% of rendering time (450ms on top of 1900ms)
+
+3. Style updates are expensive
+
+   The slowest part of the current stack is updating styles based on props or state. Big changes take linearly longer, worse on slower devices.
+
+   Benchmarks:
+
+   i. Lab benchmarks: updating a page with 1000 components, static runtime css vs. dynamic runtime css (thanks @mattcosta7 for these!)
+
+   re-rendering components with fixed styles on runtime: 242ms
+   re-rendering components with dynamic styles on runtime: 441ms (55% slower)
+
+   ii. Application benchmarks: opting out of IconButton for PR diff view (>500 files, [gh/gh#243269](https://github.com/github/github/pull/243269))
+
+   dynamic styles with sx prop: 400ms
+   dynamic styles with .css files: 165ms (60% faster)
+
+The other reasons for change are to utilise css variables for theming and improve the guardrails of consuming primer/primitives
+
+## Decisions
+
+1. Manually refactor styles to css files + css modules
+
+   Use css variables from primer/primitives for inexpensive theming support, especially with server side rendering
```

Comment:
What's the expectation for using CSS Custom Properties along with stuff like breakpoints? From what I remember, syntax like the following would not work:

```css
@media (max-width: var(--some-breakpoint)) {
  // ...
}
```

Since the `@media` query is not scoped under a `:root`. Would we use helper mixins/functions/etc?

Reply:
We ship CSS vars to work with the custom media plugin 🎉

Reply:
@siddharthkp what would the build step for this end up looking like? Would it be at publish/build time for the package or are we expecting end users to include this as part of their build toolchain?

@langermank could you share an example of how the CSS Custom Properties from primitives would interop with the custom media plugin?

Reply:
I have this running in a remix app with global CSS (not CSS modules). I have a watcher that compiles the PostCSS to CSS.

Media query looks like this
```
@custom-media --md-n-above (width >= 768px);
```

PostCSS file
```
@media (--md-n-above) {
  .AppLayout {
    --sidepane-width: var(--size-280);
    --header-height: var(--size-64);
    --topbar-visibility: visible;
  }
}
```

Built CSS file

```
@media (min-width: 768px) {
  .AppLayout {
    --sidepane-width: var(--size-280);
    --header-height: var(--size-64);
    --topbar-visibility: visible;
  }
}
```

Reply:
Thanks @langermank! Appreciate it 🙏

Reply:
If it helps the discussion at all, I'm a big fan of what Bootstrap does with their breakpoints
https://getbootstrap.com/docs/5.0/layout/breakpoints/